diff --git a/src/Components/Web.JS/src/webpack.config.js b/src/Components/Web.JS/src/webpack.config.js
index b0651fa75c..d5e4dab5de 100644
--- a/src/Components/Web.JS/src/webpack.config.js
+++ b/src/Components/Web.JS/src/webpack.config.js
@@ -26,7 +26,7 @@ module.exports = (env, args) => ({
         providedExports: true,
         usedExports: true,
         innerGraph: true,
-        minimize: true,
+        minimize: false,
         minimizer: [new TerserJsPlugin({
             terserOptions: {
                 ecma: 2019,
