diff --git a/src/Components/Web.JS/src/Rendering/BrowserRenderer.ts b/src/Components/Web.JS/src/Rendering/BrowserRenderer.ts
index 7496dd5522..30c0455266 100644
--- a/src/Components/Web.JS/src/Rendering/BrowserRenderer.ts
+++ b/src/Components/Web.JS/src/Rendering/BrowserRenderer.ts
@@ -5,7 +5,8 @@ import { RenderBatch, ArrayBuilderSegment, RenderTreeEdit, RenderTreeFrame, Edit
 import { EventDelegator } from './Events/EventDelegator';
 import { LogicalElement, PermutationListEntry, toLogicalElement, insertLogicalChild, removeLogicalChild, getLogicalParent, getLogicalChild, createAndInsertLogicalContainer, isSvgElement, permuteLogicalChildren, getClosestDomElement, emptyLogicalElement, getLogicalChildrenArray, depthFirstNodeTreeTraversal } from './LogicalElements';
 import { applyCaptureIdToElement } from './ElementReferenceCapture';
-import { attachToEventDelegator as attachNavigationManagerToEventDelegator } from '../Services/NavigationManager';
+// B.W.S: Import functions needed for detaching the navigation manager from the event delegator.
+import { attachToEventDelegator as attachNavigationManagerToEventDelegator, detachFromEventDelegator as detachNavigationManagerFromEventDelegator } from '../Services/NavigationManager';
 import { applyAnyDeferredValue, tryApplySpecialProperty } from './DomSpecialPropertyUtil';
 const sharedTemplateElemForParsing = document.createElement('template');
 const sharedSvgElemForParsing = document.createElementNS('http://www.w3.org/2000/svg', 'g');
@@ -23,15 +24,66 @@ export class BrowserRenderer {
 
   private childComponentLocations: { [componentId: number]: LogicalElement } = {};
 
+  // B.W.S: When a Blazor micro-frontend is unmounted from the page by single-spa, it needs to be
+  // able to completely dispose of all its Razor components before potentially a different Blazor
+  // micro-frontend boots and mounts itself. This is because Blazor relies on different pieces of
+  // state that are installed on the browser's global window object. If the second Blazor
+  // micro-frontend begins to boot and overwrites some of that global state (e.g., references to the
+  // Blazor or DotNet objects), then the first Blazor micro-frontend might end up calling into the
+  // second micro-frontend's .NET runtime during component disposal. This will lead to errors and
+  // ultimately an inconsistent state that we want to avoid.
+  //
+  // The intent is that this promise will be initialized when the first component is added to this
+  // BrowserRenderer. Once all root components have been disposed and child component locations have
+  // been cleared, this promise will be resolved. Then inside the Blazor micro-frontend's unmount
+  // lifecycle hook, it can await this promise either directly or (more likely) indirectly via the
+  // ensureRazorComponentsDisposed function exposed on the Blazor object.
+  //
+  // There are potentially some edge cases where this particular implementation does not work. In
+  // particular, if there are any components that are added to this BrowserRenderer after the
+  // promise has been resolved while the Blazor micro-frontend is still awaiting it, then the
+  // micro-frontend could complete its unmount while it still had components on the page. However,
+  // it seems like in the context of a Blazor micro-frontend, this would be very peculiar behavior
+  // and likely indicate a logic bug. Thus, we don't currently try to cover this specific case.
+  private componentsDisposedPromise = Promise.resolve();
+
+  private resolveComponentsDisposedPromise: (() => void) | null = null;
+
   public constructor(browserRendererId: number) {
     this.eventDelegator = new EventDelegator(browserRendererId);
 
+    // B.W.S: The original implementation directly called attachNavigationManagerToEventDelegator.
+    // This is now factored to a separate method that is called here. This allows the navigation
+    // manager to be reattached when a Blazor micro-frontend is remounted to the page by single-spa.
+    this.attachNavigationManager();
+  }
+
+  // B.W.S: This method allows a Blazor micro-frontend to reattach the navigation manager to the
+  // renderer's event delegator when the Blazor application is remounted to the page by single-spa.
+  // This will automatically be called when Blazor is first started, specifically when the
+  // BrowserRenderer is first created. Thus, a Blazor micro-frontend should NOT call this method
+  // directly on its first mount. On all subsequent mounts, the micro-frontend should ensure this
+  // method is called, typically via the ensureDomListenersReady function on the Blazor object.
+  public attachNavigationManager(): void {
     // We don't yet know whether or not navigation interception will be enabled, but in case it will be,
     // we wire up the navigation manager to the event delegator so it has the option to participate
     // in the synthetic event bubbling process later
     attachNavigationManagerToEventDelegator(this.eventDelegator);
   }
 
+  // B.W.S: The inverse of detachNavigationManager. This is intended to be called when a Blazor
+  // micro-frontend is unmounted from the page by single-spa. The micro-frontend typically ensures
+  // this via the ensureDomListenersRemoved function on the Blazor object.
+  public detachNavigationManager(): void {
+    detachNavigationManagerFromEventDelegator(this.eventDelegator);
+  }
+
+  // B.W.S: Exposes the componentsDisposedPromise to allow, e.g., ensureRazorComponentsDisposed to
+  // await it.
+  public get componentsDisposed(): Promise<void> {
+    return this.componentsDisposedPromise;
+  }
+
   public getRootComponentCount(): number {
     return this.rootComponentIds.size;
   }
@@ -48,6 +100,34 @@ export class BrowserRenderer {
       element = createAndInsertLogicalContainer(element, indexAfterLastChild);
     }
 
+    // B.W.S: When this renderer goes from managing zero components to at least one component, we
+    // want to re-initialize our promise used for signaling when those components are disposed and
+    // no longer under this renderer's management.
+    //
+    // The renderer keeps track of the components under its management via its
+    // childComponentLocations array. The only time values are added to it are when its
+    // attachComponentToElement method is called. Generally speaking, this method appears to be
+    // called in two main scenarios:
+    //
+    // 1. When this renderer attaches a root component to a logical element.
+    // 2. When this renderer inserts a component into another component already under its management
+    //    during a render update.
+    //
+    // The second scenario is not an appropriate time to re-initialize the promise, because the
+    // component being updated is already under the management of this renderer. Thus, we only want
+    // to re-initialize the promise when we go from zero to one root components, which is only
+    // possible here.
+    //
+    // NOTE: There is intentional asymmetry when determining when to resolve this promise. Blazor
+    // will call BrowserRenderer.disposeComponent first on the root component before calling it on
+    // any of the root component's children. Thus, when we determine when to resolve this promise,
+    // we look at the actual number of entries in childComponentLocations.
+    if (this.getRootComponentCount() === 0) {
+      this.componentsDisposedPromise = new Promise(resolve => {
+        this.resolveComponentsDisposedPromise = resolve;
+      });
+    }
+
     markAsInteractiveRootComponentElement(element, true);
     this.attachComponentToElement(componentId, element);
     this.rootComponentIds.add(componentId);
@@ -100,6 +180,18 @@ export class BrowserRenderer {
     }
 
     delete this.childComponentLocations[componentId];
+
+    // B.W.S: When this renderer goes from managing at least one component to zero components, we
+    // want to resolve our promise used for signaling when those components are disposed and no
+    // longer under this renderer's management.
+    //
+    // NOTE: There is intentional asymmetry when determining when to resolve this promise. Blazor
+    // will call BrowserRenderer.disposeComponent first on the root component before calling it on
+    // any of the root component's children. Thus, when we determine when to resolve this promise,
+    // we look at the actual number of entries in childComponentLocations.
+    if (Object.keys(this.childComponentLocations).length === 0) {
+      this.resolveComponentsDisposedPromise?.();
+    }
   }
 
   public disposeEventHandler(eventHandlerId: number): void {
