diff --git a/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts b/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts
index 6bf151259e..7268fe1859 100644
--- a/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts
+++ b/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts
@@ -14,5 +14,7 @@ import { BINDINGType, MONOType } from 'dotnet/dotnet-legacy';
 import { fetchAndInvokeInitializers } from '../../JSInitializers/JSInitializers.WebAssembly';
 import { JSInitializer } from '../../JSInitializers/JSInitializers';
+// B.W.S: Import the function used for getting the configured WebAssembly options.
+import { getWebAssemblyOptions } from '../../Boot.WebAssembly.Common';
 
 // initially undefined and only fully initialized after createEmscriptenModuleInstance()
 export let dispatcher: DotNet.ICallDispatcher = undefined as any;
@@ -122,17 +122,26 @@ async function importDotnetJs(startOptions: Partial<WebAssemblyStartOptions>): P
+  // B.W.S: For a Blazor micro-frontend, we would typically expect the URL to the dotnet.js be
+  // provided by a custom boot resource loader. However, in case one was not provided, then we would
+  // expect the script to be available from the asset base URL. We default to the page's base URL in
+  // order to preserve Blazor's default behavior as much as possible.
+  const baseUrl = startOptions.assetBaseUrl ?? new URL(document.baseURI);
+
   // Allow overriding the URI from which the dotnet.*.js file is loaded
   if (startOptions.loadBootResource) {
     const resourceType: WebAssemblyBootResourceType = 'dotnetjs';
     const customSrc = startOptions.loadBootResource(resourceType, 'dotnet.js', '_framework/dotnet.js', '', 'js-module-dotnet');
     if (typeof (customSrc) === 'string') {
-      const absoluteSrc = (new URL(customSrc, document.baseURI)).toString();
+      // B.W.S.: Adjust the source URL by our chosen base.
+      const absoluteSrc = (new URL(customSrc, baseUrl)).toString();
       return await import(/* webpackIgnore: true */ absoluteSrc);
     } else if (customSrc) {
       // Since we must load this via a import, it's only valid to supply a URI (and not a Request, say)
       throw new Error(`For a ${resourceType} resource, custom loaders must supply a URI string.`);
     }
   }

+  // B.W.S.: Adjust the source URL by our chosen base.
+  const absoluteSrc = (new URL('dotnet.js', baseUrl)).toString();
   // @ts-ignore: This dynamic import is handled at runtime and does not need a type declaration.
-  return await import(/* webpackIgnore: true */ "./dotnet.js");
+  return await import(/* webpackIgnore: true */ absoluteSrc);
 }
 
@@ -209,6 +220,10 @@ async function configureRuntimeInstance(): Promise<PlatformApi> {
 
   attachDebuggerHotkey(getConfig());
 
+  // B.W.S: Patch the runtime's internal dynamic_import API so that the correct asset base URL is
+  // applied to imports from absolute and relative URL paths.
+  patchRuntimeDynamicImport(mono_internal);
+
   Blazor.runtime = runtime;
   Blazor._internal.dotNetCriticalError = printErr;
   setModuleImports('blazor-internal', {
@@ -227,6 +242,41 @@ async function configureRuntimeInstance(): Promise<PlatformApi> {
   };
 }
 
+// B.W.S: Patch the dynamic_import function from the .NET browser runtime's internal API. We do this
+// because the Blazor WebAssemblyHost (which resides in the C# side of Blazor) will import the .NET
+// hot reload JavaScript module via JS interop. It specifies the location of the script via the
+// absolute URL path "/_framework/hot-reload.js".
+//
+// This interop call will eventually make its way to the runtime's dynamic_import function. This
+// function is defined at the .NET runtime layer in .NET 8+. Prior to .NET 8, the import was handled
+// by the Microsoft.JSInterop.JS module that was bundled with the Blazor startup script, which we
+// were able to modify relatively easily as part of the patches B.W.S. makes to Blazor.
+//
+// However, now we can't (easily) modify code defined by the runtime. Instead, we define an
+// interceptor for the runtime's dynamic_import function that will inspect the given module URL. If
+// it is a relative or absolute URL path, we apply the asset base URL (if available) from the
+// WebAssembly options. If an asset base URL is not available, we pass the module URL unchanged to
+// the runtime's dynamic_import function.
+function patchRuntimeDynamicImport(internalRuntime: any) {
+  const webAssemblyOptions = getWebAssemblyOptions();
+  const assetBaseUrl = webAssemblyOptions.assetBaseUrl;
+  const originalDynamicImport = internalRuntime.dynamic_import;
+
+  const adjustUrlPath = (path: string) => {
+    if (path.startsWith('/')) {
+      return new URL(path.substring(1), assetBaseUrl).toString();
+    } else if (path.startsWith('./')) {
+      return new URL(path, assetBaseUrl).toString();
+    }
+    return path;
+  };
+
+  internalRuntime.dynamic_import = (moduleName: string, moduleUrl: string): Promise<any> => {
+    moduleUrl = adjustUrlPath(moduleUrl);
+    return originalDynamicImport(moduleName, moduleUrl);
+  };
+}
+
 function setProgress(resourcesLoaded, totalResources) {
   const percentage = resourcesLoaded / totalResources * 100;
   document.documentElement.style.setProperty('--blazor-load-percentage', `${percentage}%`);
